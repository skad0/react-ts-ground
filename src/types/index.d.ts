export interface IVideo {
    url: string;
}

export interface IApplicationState {
    currentVideo: IVideo|null;
    videoList: IVideo[];
}
