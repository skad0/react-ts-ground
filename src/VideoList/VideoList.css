.VideoList {
    padding: 15px 0;
}

.VideoList-List {
    margin: 0;
    padding: 0;
    list-style: none;
}

.VideoList-Item {
    height: 30px;

    padding: 5px;

    box-sizing: border-box;

    border-bottom: 1px dashed #aaa;

    cursor: pointer;

    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
}

.VideoList-Item_current {
    background: #ffdb4d;
}

.VideoList-RemoveItem {
    background: transparent;
    border: 0;
}
