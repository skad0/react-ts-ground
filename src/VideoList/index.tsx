import * as React from 'react';

import {IVideo} from '../types';

import './VideoList.css';

interface IProps {
    list: IVideo[];
    handleChoose?: (video: IVideo) => void;
}

class VideoList extends React.Component<IProps> {
    public render() {
        const {list} = this.props;

        return (
            <div className="VideoList">
                {
                    list.length > 0 ?
                        <ul className="VideoList-List">
                            {this.renderListOptions.call(this, list)}
                        </ul> :
                        'No videos provided yet'
                }
            </div>
        );
    }

    private renderListOptions(list: IVideo[]):JSX.Element[] {
        return list.map((video, index) => {
            return <li key={index}
                    onClick={this.onVideoClick.bind(this, video)}
                    className={`VideoList-Item ${video.current ? 'VideoList-Item_current' : ''}`}>
                {video.url}
            </li>;
        });
    }

    private onVideoClick(video: IVideo):void {
        const {handleChoose} = this.props;

        if (!handleChoose) {
            return;
        }

        handleChoose(video);
    }
}

export default VideoList;
