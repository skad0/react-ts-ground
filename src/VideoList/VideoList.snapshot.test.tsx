import * as React from 'react';
import * as renderer from 'react-test-renderer';

import VideoList from './index';

import {IVideo} from '../types';

const video: IVideo = {
    id: 1,
    url: 'https://www.youtube.com/watch?v=OkwplGEUhkM'
};

describe('VideoList', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<VideoList list={[video]} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with active video', () => {
        const tree = renderer
            .create(<VideoList list={[Object.assign({current: true}, video)]} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('render message if empty list provided', () => {
        const tree = renderer
            .create(<VideoList list={[]} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

