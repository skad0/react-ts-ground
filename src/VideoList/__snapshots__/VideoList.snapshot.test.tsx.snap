// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoList render message if empty list provided 1`] = `
<div
  className="VideoList"
>
  No videos provided yet
</div>
`;

exports[`VideoList renders correctly 1`] = `
<div
  className="VideoList"
>
  <ul
    className="VideoList-List"
  >
    <li
      className="VideoList-Item "
      onClick={[Function]}
    >
      https://www.youtube.com/watch?v=OkwplGEUhkM
    </li>
  </ul>
</div>
`;

exports[`VideoList renders correctly with active video 1`] = `
<div
  className="VideoList"
>
  <ul
    className="VideoList-List"
  >
    <li
      className="VideoList-Item VideoList-Item_current"
      onClick={[Function]}
    >
      https://www.youtube.com/watch?v=OkwplGEUhkM
    </li>
  </ul>
</div>
`;

exports[`VideoList renders correctly with remove provided 1`] = `
<div
  className="VideoList"
>
  <ul
    className="VideoList-List"
  >
    <li
      className="VideoList-Item "
      onClick={[Function]}
    >
      <button
        className="VideoList-RemoveItem"
      >
        X
      </button>
      https://www.youtube.com/watch?v=OkwplGEUhkM
    </li>
  </ul>
</div>
`;
